server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -380,15 +381,24 @@ private static boolean upgradeLegacyNodeFolders(Logger logger, Settings settings
 
                 // determine folders to move and check that there are no extra files/folders
                 final Set<String> folderNames = new HashSet<>();
+                final Set<String> expectedFolderNames = new HashSet<>(Arrays.asList(
+
+                    // node state directory, also containing MetaDataStateFormat-based global metadata
+                    MetaDataStateFormat.STATE_DIR_NAME,
+
+                    // Lucene-based metadata folder
+                    LucenePersistedStateFactory.METADATA_DIRECTORY_NAME,
```

Comment:
I don't really see much difference either way, but I personally prefer the conceptual separation here. On dedicated master nodes we could indeed drop the node metadata file in due course. We avoid sharing a single folder between a Lucene index and some MetaDataStateFormat-based files in other places but we could break that pattern here if needed. I think repurposing will be a little simpler with separate folders, if only because there's no need to implement a new delete operation on `MetaDataStateFormat` when you can just wipe out the whole directory.